### ISSUE TITLE
Week 2 Project: News Reader Powered by the New York Times

### ISSUE THREAD

**Issue by smacgregor** [smacgregor/NewYorkTimesSearch#14]
My app is complete, please review. /cc @codepathreview @codepath

This was a fun project. I particularly enjoyed working with RecyclerViews / staggered grid layouts. 

<h2>Interesting Notes</h2>

- Leveraged the fact that enums are classes in Java to [data drive processing of search filter criteria](https://github.com/smacgregor/NewYorkTimesSearch/blob/master/app/src/main/java/com/smacgregor/newyorktimessearch/searching/SearchFilter.java#L43). This made it easy to add an extra sort order (relevance) and an additional news desk (like Foreign)
- I like Picasso more than Glide - maybe I'm just biased by the smaller library size.
- Instead of picking a random thumbnail for the staggered grid layout - I first [filtered out small thumbnail images](https://github.com/smacgregor/NewYorkTimesSearch/blob/master/app/src/main/java/com/smacgregor/newyorktimessearch/core/ArticleHelpers.java#L21)  as they are too small to look good on a high res mobile device. Thought this looked better overall. 
- The RecyclerView Animators library is really sweet. I had fun trying out different adapter animations. 

<h2>Open Questions</h2>

1. Use of anonymous functions. I tend to keep methods in classes flat instead of using anonymous methods inside of other methods. 

Instead of doing this:

```
searchResultsView.addOnScrollListener(new EndlessRecyclerViewScrollListener(linearLayoutManager) {           @Override 
public void onLoadMore(int page, int totalItemsCount) {         
       loadMoreSearchResults(page);     
} });
```

I tend to have:

```
    @Override
    public void onItemClick(View view, int position) {
        *snip*
    }

    @Override
    public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
        *snip*
    }

    @Override
    public boolean onQueryTextSubmit(String query) {
       *snip*
    }
```

Is one style preferred over the other in Android? I find the flat list of methods easier to read.
1. Can we bind elements in onCreateOptionsMenu? Butterknife crashed for me here:

```
@Bind(R.id.menu_item_search) MenuItem mSearchMenuItem;
```

I understand why it crashed and I don't think there's a way to make it word but wanted to ask.
1. It was a little frustrating that in order to use the Parceler library I had to make my model's member variables public.
2. Using Glide I found I still needed to use a DynamicHeighImageView element to get the resizing to work right. One of the advantages of Glide is how it handles resizing. Should I have been able to optimize away my use of DynamicHeighImageView with Glide?

<h2>From the if I had more time department</h2>

- Wish I had time to investigate how to use https://github.com/afollestad/material-dialogs to improve styling of the search filter fragment
- Progress spinners inside the RecyclerView when loading more results
- Progress spinner when searching for results

**Comment by codepathreview** [comment]
Hi @smacgregor,

Here are some inputs for your questions:
1. Both the styles are fine depending on your implementation. Different libraries have different implementations. 
2. Don't think MenuItems are supported by Butterknife : [See issue here](https://github.com/JakeWharton/butterknife/issues/41) , Jake Wharton commented that he wants to keep Butterknife  "focused on assisting with view-realted things only."
3. Agree, but some libraries have their own limitations. I personally like to write the parcels myself. [Follow this guide.](https://guides.codepath.com/android/using-parcelable)
4. Not really sure about that, need to investigate more.

**Comment by codepathreview** [comment]
:+1: Excellent work. A few notes after checking out the code:
- Consider [better organizing your source files](http://guides.codepath.com/android/Organizing-your-Source-Files)
- Properly reused the single base method in search activity to fetch results for both initial load and the pagination
- Nice to see you added the search box to the ActionBar!
- Properly used checkboxes/spinners (with populated values) to allow users to select filters
- Properly sent and received relevant filters using a bundle to pass data between activities
- Good to see you used a filters object to represent the filters to apply to the query
-  Nice job adding the share intent to the article detail view!
- Consider adding [progress bars](http://guides.codepath.com/android/Handling-ProgressBars#progress-within-actionbar) to your app whenever a network request goes out

Also, here's a more detailed [Project 2 Feedback Guide here](http://courses.codepath.com/snippets/intro_to_android/project_2_feedback) which covers the most common issues with this submitted project. Read through the [feedback guide point-by-point](http://courses.codepath.com/snippets/intro_to_android/project_2_feedback) to determine how you could improve your submission.

Let us know if you have any other thoughts or questions about this assignment. The next assignment (Twitter Client) will be especially important since it introduces the majority of the remaining pieces necessary to build a fully functional API client with complex feeds of data and user creation.

**Comment by smacgregor** [comment]
On an unrelate note I'm really liking the styling and naming convention guide used by ribot:

https://github.com/ribot/android-guidelines/blob/master/project_and_code_guidelines.md

**Comment by smacgregor** [comment]
> Properly reused the single base method in search activity to fetch results for both initial load and the pagination

Yeah I started down that path but then felt the error messages  / the message when there are no matches and the behaviors were slightly different enough to warrant separate methods for code readablity.  I'll think about collapsing them back into one. Thanks for the suggestion.

> Consider better organizing your source files

I might need a little more guidance on this as this was similar to feedback I received from the first assignment as well. I've chosen to organize by feature and broke things down into the following packages:

1) core - foundational pieces like our article and thumbnail models that are used throughout the app.
2) viewing - all activities and models related to viewing an article
3) searching - all activities and models related to searching for articles including the main activity - SearchActivity.java
4) networking - our article data provider client

**Comment by codepathreview** [comment]
In general, it should be based on the component of android
- activities - which has all activities
- services - for all services (if any)
- receivers - all broadcast receivers
- adapters - all adapters 
- views - custom views 
- fragments - all fragments belong here
- util/helper - where you put `Utils.java`, `LayoutUtils.java` etc
- interfaces
- network - REST clients, etc

Hope you got the idea?
